File: 29Items-Projects/7Algorithms/6-Backtracking/java/src/ratinmaze/RatInMaze.java
package ratinmaze;

import java.util.*;

public class RatInMaze {

    private final int[][] maze;
    private final int n;
    private final List<int[][]> solutions = new ArrayList<>();

    private static final int[] DR = {0, 1, 0, -1};
    private static final int[] DC = {1, 0, -1, 0};

    public RatInMaze(int[][] maze) {
        if (maze == null || maze.length == 0)
            throw new IllegalArgumentException("Maze must not be empty");
        this.maze = maze;
        this.n = maze.length;
    }

    public void solveAndPrint() {
        int[][] path = new int[n][n];
        if (maze[0][0] == 1) {
            path[0][0] = 1;
            backtrack(path, 0, 0);
        }

        System.out.printf("Found %d path(s) in %dx%d maze%n%n", solutions.size(), n, n);

        int limit = Math.min(solutions.size(), 4);
        for (int i = 0; i < limit; i++) {
            System.out.printf("Path #%d:%n", i + 1);
            printPath(solutions.get(i));
            System.out.println();
        }
        if (solutions.size() > limit) {
            System.out.printf("... and %d more.%n", solutions.size() - limit);
        }
    }

    private void backtrack(int[][] path, int row, int col) {
        if (row == n - 1 && col == n - 1) {
            solutions.add(copyPath(path));
            return;
        }
        for (int d = 0; d < 4; d++) {
            int nr = row + DR[d], nc = col + DC[d];
            if (isSafe(path, nr, nc)) {
                path[nr][nc] = 1;
                backtrack(path, nr, nc);
                path[nr][nc] = 0;
            }
        }
    }

    private boolean isSafe(int[][] path, int r, int c) {
        return r >= 0 && r < n && c >= 0 && c < n
                && maze[r][c] == 1 && path[r][c] == 0;
    }

    private int[][] copyPath(int[][] path) {
        int[][] copy = new int[n][n];
        for (int i = 0; i < n; i++) copy[i] = path[i].clone();
        return copy;
    }

    private void printPath(int[][] path) {
        for (int r = 0; r < n; r++) {
            StringBuilder sb = new StringBuilder("|");
            for (int c = 0; c < n; c++) {
                if (maze[r][c] == 0) {
                    sb.append(" # ");
                } else if (path[r][c] == 1) {
                    sb.append(" * ");
                } else {
                    sb.append(" . ");
                }
            }
            sb.append("|");
            System.out.println(sb);
        }
    }
}

File: 29Items-Projects/7Algorithms/6-Backtracking/python/rat_in_maze.py
"""Rat in a Maze — find path through a maze grid from (0,0) to (n-1,n-1)."""

_DIRS = [(0, 1), (1, 0), (0, -1), (-1, 0)]  # right, down, left, up


class RatInMaze:
    def __init__(self, maze: list[list[int]]):
        if not maze:
            raise ValueError("Maze must not be empty")
        self.maze = maze
        self.n = len(maze)
        self.solutions: list[list[list[int]]] = []

    def solve_and_print(self) -> None:
        path = [[0] * self.n for _ in range(self.n)]
        if self.maze[0][0] == 1:
            path[0][0] = 1
            self._backtrack(path, 0, 0)

        print(f"Found {len(self.solutions)} path(s) in {self.n}x{self.n} maze\n")

        limit = min(len(self.solutions), 4)
        for i, sol in enumerate(self.solutions[:limit], 1):
            print(f"Path #{i}:")
            self._print_path(sol)
            print()
        if len(self.solutions) > limit:
            print(f"... and {len(self.solutions) - limit} more.")

    def _backtrack(self, path: list[list[int]], row: int, col: int) -> None:
        if row == self.n - 1 and col == self.n - 1:
            self.solutions.append([r[:] for r in path])
            return
        for dr, dc in _DIRS:
            nr, nc = row + dr, col + dc
            if self._is_safe(path, nr, nc):
                path[nr][nc] = 1
                self._backtrack(path, nr, nc)
                path[nr][nc] = 0

    def _is_safe(self, path: list[list[int]], r: int, c: int) -> bool:
        return 0 <= r < self.n and 0 <= c < self.n and self.maze[r][c] == 1 and path[r][c] == 0

    def _print_path(self, path: list[list[int]]) -> None:
        for r in range(self.n):
            row = "|"
            for c in range(self.n):
                if self.maze[r][c] == 0:
                    row += " # "
                elif path[r][c] == 1:
                    row += " * "
                else:
                    row += " . "
            row += "|"
            print(row)
